Fix "Rate exceeded" issues for large clusters

mattcallanan opened the pull request:
Many fixes and improvements - mainly addressing timeouts and AWS API rate limit issues for large clusters.
Client now uses short polling to request cluster data from server. Server returns immediately and continues processing in the background. Server caches cluster data for 30mins with a "forceRefresh" option available to client (exposed as "Reload Server Cache" button). Also externalised server-side defaults into environment-based configuration files.

UI:
- Implemented short polling logic - query the server up to 120 times with 5s intervals
- Added "Reload Server Cache" button to UI to force server to refresh cluster data.
- Upgraded bootstrap from v3.3.5 to v3.3.7 for glyphicons including "glyphicons-halflings-regular" fonts.
- Improved error handling.
- Fix 'onError is not a function' bug.
- Fix displaying of server errors. "TypeError: Cannot read property 'length' of undefined at renderGraph (graph.js:103)"
- Handle errors caught before 'graph' object is initialised without causing "Uncaught TypeError: Cannot read property 'append' of undefined" error for graph.
- Show instances even if there are no tasks.

Server:
- Externalised server-side defaults into environment-based configuration files.
- Added ClusterState & ClusterStateCache to cache cluster data.
- Added FetchStatus to represent current status of fetching cluster data.
- Improved error handling.
- Added test suite - run tests with "npm run test"
- Upgrade Dockerfile base image to node:9.11.1-alpine with fix for "gyp": "gyp ERR! stack Error: Can't find Python executable "python", you can set the PYTHON env variable"
- Added delayPromise() to introduce a 1s delay between ECS API calls to avoid "Rate exceeded" back-pressure issues when running within AWS account.
- Fix "InvalidParameterException: instanceIds can have at most 100 items" error.
- Use internal DEBUG_LOGGING constant to assist when debugging locally
- Use template strings and replace some "var"s with "let"s

Commit:
Fix "Rate exceeded" issues for large clusters.
- Dockerfile tweaks:
  - Upgrade docker base image to node:8.5-alpine with fix for "gyp": "gyp ERR! stack Error: Can't find Python executable "python", you can set the PYTHON env variable"
- Fixes and tweaks to index.js:
  - Added delayPromise() to introduce a 1s delay between ECS API calls to avoid "Rate exceeded" back-pressure issues when running within AWS account.
  - Fix "InvalidParameterException: instanceIds can have at most 100 items" error.
  - Use internal DEBUG_LOGGING constant to assist when debugging locally
  - Use template strings and replace some "var"s with "let"s
- UI Fixes:
  - Fix 'onError is not a function' bug.
  - Fix displaying of server errors. "TypeError: Cannot read property 'length' of undefined at renderGraph (graph.js:103)"
  - Handle errors caught before 'graph' object is initialised without causing "Uncaught TypeError: Cannot read property 'append' of undefined" error for graph.
  - Show instances even if there are no tasks.
- Added TODO: "Make delay between AWS API calls configurable"

karlringrose commented on public/javascripts/graph.js:
```diff
@@ -92,48 +92,57 @@ function remainingResource(d, resourceType) {
+function recreateMainGraphElement(chartDivId, graphWidth, leftMargin, rightMargin, totalHeight, topMargin, bottomMargin) {
     d3.select('#' + chartDivId).select("svg").remove();
-    var graph = d3.select('#' + chartDivId)
+    let graph = d3.select('#' + chartDivId)
```
As you aren't re-assigning graph, I would recommend using const.

Commit:
Introduce server-side cache for Task Definitions.  Update to use ES6 let/const in place of var.

Commit:
Prevent client-side null error before graph is created.  Update to use ES6 let/const in place of var.

Commit:
Add README instructions for debug logging.

mattcallanan committed:
Introduce short polling approach to avoid client timeouts:
- Server responds immediately with 'fetching' state and populates cluster store asynchronously
- Client polls until cluster is in 'fetched' state

mattcallanan committed:
Fix Y axis units to show 'k' instead of 'G' when showing CPU resource

mattcallanan committed:
Fix registeredCpu values in static demo cluster files.

mattcallanan committed:
Improved short polling. Cached cluster data for 30mins with "forceRefresh" option. Externalised server-side defaults into environment-based configuration files.

UI:
- Implemented short polling logic - query the server up to 120 times with 5s intervals
- Added "Reload Server Cache" button to UI to force server to refresh cluster data.
- Upgraded bootstrap from v3.3.5 to v3.3.7 for glyphicons including "glyphicons-halflings-regular" fonts.
- Improved error handling.

Server:
- Externalised server-side defaults into environment-based configuration files.
- Added ClusterState & ClusterStateCache to cache cluster data.
- Added FetchStatus to represent current status of fetching cluster data.
- Improved error handling.
- Added test suite - run tests with "npm run test"

mattcallanan committed:
Merge branch 'master' into fix_rate_limit_issue6

# Conflicts:
#	package-lock.json

mattcallanan committed:
Merge from master.

mattcallanan committed:
Upgraded packages per 'npm audit' recommendations.

mattcallanan committed:
Upgrade Dockerfile base image from node:8.5 to node:9.11.1

mattcallanan committed:
Tweak position of timestamp on UI.

mattcallanan committed:
Updated documentation.

mattcallanan committed:
Updated documentation.

mattcallanan committed:
Refactoring. Introduce PromiseDelayer class. Rename tests from "test*.js" to "*Spec.js".

mattcallanan committed:
Use 'TARGET_ENV' as environment config loading switch (as 'NODE_ENV' is used by Express for other purposes).
